### DISCUSSION TITLE
Automatically generate IEquatable, IEqualityComparer, IComparable, toString for "simple" classes.

### DISCUSSION THREAD

**Discussion by be5invis** [dotnet/csharplang#550] [Language Ideas] [answered]
Like Haskell’s `deriving`.

**Comment by orthoxerox** [comment]
That's planned for records. Well, not `IComparable`, only equality.

**Comment by be5invis** [comment]
O i forgot `IComparer`.

**Comment by lachbaer** [comment]
Does there need to be syntax for this? Templates, code snippets and right clicking the interface name in the class declaration already do a great job. When you want to hide the implementation, just put them in a collapsible `region`.

**Comment by Richiban** [comment]
@lachbaer 

> Templates, code snippets and right clicking

I'm afraid that's not acceptable at all. Since the problem with generating the code is that the generated code risks falling out of sync with the rest of the class I'd say that hiding the implementation in a region will only exacerbate the issue.

Hopefully at least some of this will be solved when we eventually get default interface methods #406, and at least we will 'inherit' `==` and `!=` by implementing `IComparable`, and `>` and `<` by implementing `IComparable` etc.

**Comment by jnm2** [comment]
> the generated code risks falling out of sync with the rest of the class

Good use case for a build-time analyzer, btw.

**Comment by lachbaer** [comment]
> Good use case for a build-time analyzer, btw.

I've heard that so often now "build an analyzer". It would be nice if there were some kind of IDE help for this, e.g. a GUI, an Analyzer Wizard or an easy config file that could define those analyzers without diving into building a dedicated dll by self written analyzer code.

**Comment by jnm2** [comment]
@lachbaer I agree. Plus, the templates are so far out of date it's embarrassing. https://github.com/dotnet/project-system/issues/1481

**Comment by orthoxerox** [comment]
@jnm2 you haven't tried to build a JIRA plugin recently. Their project templates are even worse. They contain not just outdated, but semantically *invalid* code.

**Comment by jnm2** [comment]
@orthoxerox I'm sorry! I'm not likely to either, now that you told me that.

**Comment by Richiban** [comment]
There are still several unanswered questions, but I thoroughly support generation of at least Equality and ToString methods by the compiler through use of an attribute, as has been approved in principal for F# https://github.com/fsharp/fslang-suggestions/issues/554

**Comment by lachbaer** [comment]
Again a good use for an analyzer. Mark every method with an attribute where the analyzer should check if all fields (maybe restricted to visibility) are somehow referenced. When you add a field you will get a warning or error and you can fix that appropriately. For standard methods an automated codefix can help you.

Having that written, I think that there should be **much** more documentation on analyzers and code fixes (what else?) on the net and in every new upcoming book on the shelves!

**Comment by YairHalberstadt** [comment, marked as answer]
I think this can be considered covered between records and source generators
